### PR TITLE
VIMC-4181: Add helper for getting data from cross_all, cross_under5, cohort_all, and cohort_under5 tables

### PR DESCRIPTION
**Pull request by r-ash**
This PR will
* Setup a test postgres db (we can use this for the other testing in the package but can do that in another PR - for now only used for testing of this function)
* Add function get_stochastic_data which can be used to query `cross_all`, `cross_under5`, `cohort_all` and `cohort_under5` tables

I think this is at a good point for a first look - I think this covers 90% of cases but would could add support for some other cases like
* Support adding generated period column calculated from some range of years i.e. through generating some CASE WHEN statements
* Support adding modelling_group to this which is found from the `stochastic_file_id`

Couple of thing specifically I was thinking about
* naming of function - is there a better collective name for the `cross_all`, `cross_under5`, `cohort_all` and `cohort_under5` tables than stochastic?
* Avoided using dbplyr sql generation stuff just because it would be a big dependency to drag in - I wonder if it would make sense to use that though if we see people wanting to do anything more complex than is currently supported
* At the moment the signature takes an existing connection - should the function create this connection for the user? Thinking about the context that we will use this function that connection probably already exists

### CODE REVIEW

**Review by xiangli313 — APPROVED** [review]

**Comment by richfitz** [comment]
Summarising our discussion on this last night:

> * naming of function - is there a better collective name for the cross_all, cross_under5, cohort_all and cohort_under5 tables than stochastic?

If we change this, then `stochastic_summary` would be my suggestion? But I think this is the main entrypoint into the stochastic data at the moment so that's fine. Given there's a tight coupling between annex's state and this package we might need to do some version fetching at some point. So, let's leave it as-is for now and see if that causes problems as we use it

> * Avoided using dbplyr sql generation stuff just because it would be a big dependency to drag in - I wonder if it would make sense to use that though if we see people wanting to do anything more complex than is currently supported

As discussed, let's leave as-is now, but pull in dbplyr later if more complex SQL generation is needed, or if we create a common interface that works with connections and data.frames equally. The size of a dependency is not a great issue for this package as it is unlikely to be built on and stands more as a reference implementation.

> * At the moment the signature takes an existing connection - should the function create this connection for the user? Thinking about the context that we will use this function that connection probably already exists

I think this is the correct way to do it

File: R/stochastic.R [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#' @param annex Connection to annex db
+#' @param table One of cross_all, cross_under5, cohort_all or cohort_under5
+#' @param groups Categories to group by for aggregating in query
+#' @param filters Filters to apply before aggregation
+#'
+#' @return Mean, 0.025 and 0.975 quantiles for deaths_default, deaths_novac.
+#' deaths_impact, dalys_default, dalys_novac, dalys_impact with specified
+#' groupings.
+#' @export
+get_stochastic_data <- function(annex, table,
```
Comment (richfitz) [inline R/stochastic.R]: I wonder if `fetch_stochastic_data` might be just a touch more descriptive?

File: tests/testthat/helper-db.R [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+get_test_connection <- function() {
+  info <- get_db_info()
+  con <- get_postgres_connection(info$dbname, info$user, info$host)
+  withr::defer_parent(DBI::dbDisconnect(con))
+  con
+}
+
+vimpact_test_postgres_connection <- function(dbname, user, host) {
+  tryCatch(
+    get_postgres_connection(dbname, user, host),
```
Comment (richfitz) [inline tests/testthat/helper-db.R:34]: Great - we discussed doing this; this is perfect

File: tests/testthat/helper-db.R [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  )
+}
+
+get_postgres_connection <- function(dbname, user, host) {
+  DBI::dbConnect(RPostgres::Postgres(), dbname = dbname, user = user,
+                 host = host)
+}
+
+add_dummy_data <- function(con) {
+  message("Adding dummy data to database")
```
Comment (richfitz) [inline tests/testthat/helper-db.R]: Can you please add a check here to make sure that this is never run on annex itself? I think that you can check `con` to find out if it's localhost (even if you just `format` it and look for `127.0.0.1` or equivalent, but possibly by looking at its slots - try `slotNames(con)`).

It's unlikely but this function might accidentally do a nasty delete on annex. I presume that ordinarily we have only a readonly user though? Still `DROP TABLE...CASCADE` is a scary looking command!

**Review by richfitz — CHANGES_REQUESTED** [review]
Looks good, except I think that adding a little check before dropping tables could save a headache in the future.

The other way of fixing that would not be to pass the connection there but to create the connection to the test db right before the delete so that it can't be used on the real annex

File: tests/testthat/helper-db.R (hunk 2 above)
Comment (r-ash) [reply tests/testthat/helper-db.R:34]: That wasn't actually being called from the tests and instead being called from the pkg load which wasn't ideal. I've fixed this up now

File: tests/testthat/helper-db.R (hunk 3 above)
Comment (r-ash) [reply tests/testthat/helper-db.R]: Added a check on the db name - if it doesn't match `vimpact_test_db` then we throw an error

File: R/stochastic.R (hunk 1 above)
Comment (r-ash) [reply R/stochastic.R]: Yep sounds good

**Review by richfitz — APPROVED** [review]

**Pull request merged**